### PR TITLE
Some small-ish changes to auto-detect the loop devices used

### PR DESCRIPTION
**Pull request by mvastola**
This is needed if the system already has one or more loop devices assigned. 

* Also added in support for a loop device as the target (useful when using this script to create an installer for a VM) by running `kpartx` on the target device.
* Finally fixed an issue wherein `umount`ing of the target device will fail if it currently has more than one partition.

### CODE REVIEW

File: mkosxinstallusb.sh [hunk 1]
```diff
@@ -24,6 +24,11 @@ cleanup () {
     rm -f InstallESD.img BaseSystem.img
 }
 
+map_partitions () {
+  set +e
+  kpartx -sav "$1" | sed -r 's/^add\s+map\s+(\S+)p[0-9]+\s.*$/\/dev\/mapper\/\1/;t;d' | head -n1
```
Comment (eprigorodov) [inline mkosxinstallusb.sh:29]: When this is called for the USB device in the line 129,

> stick_partitions="$(map_partitions "$stick_dev")"

then `kpartx` responds with the USB device name, like this:

> add map sdc1 (252:3): 0 63010783 linear 8:32 2048

which does not contain '/dev/loop?p?' and thus the regexp `(\S+)p[0-9]` does not match.

Due to that `$stick_partitions` in the line 129 becomes an empty string, so `mkfs.hfsplus` in the line 130 and `mount` in the line 138 fail. But the script does not stop because `map_partitions()` had `+e` set, so it copies 5Gb+ of files into the `/mnt/OSX_installer_drive`.

Comment (eprigorodov) [reply mkosxinstallusb.sh]: The point itself is very valid. It is incorrect to hardcode a specific loop device because to the moment of the script run any number of them can be allocated.

File: mkosxinstallusb.sh [hunk 2]
```diff
@@ -107,30 +112,30 @@ fi
...
-kpartx -s -a BaseSystem.img
+BaseSystem_partitions="$(map_partitions "BaseSystem.img")"
 
 echo "\n# partitioning USB drive"
-umount $stick_dev? || true
+ls -1 "${stick_dev}"? | xargs -n1 umount || true
 sgdisk -o $stick_dev
 partprobe $stick_dev
 sgdisk -n 1:0:0 -t 1:AF00 -c 1:"disk image" -A 1:set:2 $stick_dev
-partprobe $stick_dev
```
Comment (eprigorodov) [inline mkosxinstallusb.sh:123]: `sgdisk` in the line above re-partitions USB stick to contain a single HFS+ partition. Without `partprobe` kernel would still see the old partition table, so formatting can do any kind of damage.

**Comment by eprigorodov** [comment]
The point itself is very valid. It is incorrect to hard-code a specific loop device because to the moment of the script run any number of them can be allocated.

I would just propose the following:

* Do not call `map_partitions()` for finding out the layout of USB drive. Rationale: a few lines above it was just re-partitioned by `sgdisk` to a predefined GPT with only one HFS+ partition.
* Do not remove the `partprobe` call, without it the kernel would still see the old GPT of USB stick, whereas it just has been changed.
* Simplify the `sed` regexp in the `map_partitions()`, assuming that it always use loop devices, so its output must contain `loop\d+p\d+`.
* Do not suppress errors in `map_partitions()` so it better fails early rather than does unpredictable damage.

File: mkosxinstallusb.sh (hunk 1 above)
Comment (mvastola) [reply mkosxinstallusb.sh:29]: Ok. It looks like I just have to remove the `p` on line 24. Perhaps the capture group should also be changed to `[a-z]+` to be safe. (I forget the exact rules on regex greediness).

And I can't for the life of me remember why I put the `set +e`. I think I copied it from the `cleanup` function. There also isn't even a `set -e` in the first place (though I agree there should be) which makes this even more dumb. I'll take it out and add one to the start of the script.

File: mkosxinstallusb.sh (hunk 2 above)
Comment (mvastola) [reply mkosxinstallusb.sh:123]: Are you sure on this? The reason why I removed it (and also the reason why I don't use `partprobe` everywhere in lieu of `kpartx` is that in my experience `kpartx` is a bit more standardized.
Reasons for using only `kaprtx` include:
* `kpartx` always seems to put the partitions into `/dev/mapper/{disk}p{partnum}`
* I'm not sure sure all distros create partitions for loop devices with `partprobe` (and, for the ones that do, they won't necessarily put them in `/dev/mapper`.
* In the past, I've experienced loop devices refusing to detach if partitions from `kpartx` haven't been unmapped (I think this is fixed in newer kernels and/or versions of `kpartx`). As a result, a concern of mine was that if `partprobe` does map these partitions (and if say it puts them in `/dev` and not `/dev/mapper`, the user would have issues detaching the loop device. I generally wasn't sure about the mechanics here, so I figured I'd be safe.
* I was under the impression that `kpartx` re-reads the partition table as does `partprobe`. Whenever I edit a partition table with `fdisk` and that utility can't update the partition table, it tells you to run `partprobe` *or* `kpartx`.

**Comment by mvastola** [comment]
Thanks for checking out this PR. I addressed all but the first point by replying to your comments. Please reply back to those and we'll figure them out.

Re the first bullet point, I'm not sure I understand fully. Are you simply saying it's redundant, if so, i disagree, unless we add back `partprobe` (a possibility I addressed in the inline comments). In any case, it just seemed simpler to not have two different procedures based on the type of the underlying device. Also, since `partprobe` and `kpartx` can put the block devices in different folders (`/dev` vs `/dev/mapper`, respectively) and since loop devices have the letter `p` prepended to the partition number, the code would get needlessly more complex in my opinion.  

Let me know your thoughts on all this and we can resolve these issues.

I'll go ahead and make the changes I mentioned above re the first inline comment tomorrow as I'm not currently near a computer.

File: mkosxinstallusb.sh (hunk 2 above)
Comment (eprigorodov) [reply mkosxinstallusb.sh:123]: I have checked on some old Ubuntu 14 instance and can confirm that `partprobe` is really not needed. Some years ago kernel did not re-read new partition table after making changes, so that required restarting or running `partprobe`. But now new partitions are visible right after the `sgdisk` run, so this call can be removed.

**Comment by eprigorodov** [comment]
Thank you for considering my comments :)

With regard to the first point let me rephrase the same concern in another way:

`sgdisk` re-partitions the USB drive to a fixed layout that contains a single HFS+ partition. So device name `"$stick_dev"1` will be surely correct. What is much more important, this device name will always be based on the user-supplied value for `$stick_dev`. Remember, the script does some precaution checks and warns user that everything under e.g. `/dev/sdc` will be destroyed.

The `map_partitions()` creates a **new** mapper device and then tries to detect its name. That fails now because regexp expects to parse`add map loop1p1 ...`and cannot match `add map sdc1` returned for the USB stick.
It is possible to fix the regexp, but I don't like the very idea of formatting device that was dynamically identified by using regular expressions and `sed`.

**Comment by eprigorodov** [comment]
I am fully fine about using `map_partitions()` for image files, because that is only read access and possible errors cannot bring much harm. 

Should we agree on the first point then please commit your fixes and I will merge the PR.

**Comment by mvastola** [comment]
My point though is that `"$stick_dev"1` is not sure to be correct, at least on loop devices.

In that case, at the very least it will be  `"$stick_dev"p1`, but as I mentioned, some Linux installs don't map loop device partitions to `/dev/loopNpX`. Using `kpartx` ensures consistency there so the point of this function is a one-size fits all.

So if we didn't use the regex for physical devices, you'd need to have an if/then in `map_partitions` and return the prefix without the partition number. This could be done by simply returning the argument to `map_partitons` if it doesn't contain `loop`, but it just seemed simpler to do everything the same way. 

If you want though, I'll change it. Just didn't see the point.

**Comment by eprigorodov** [comment]
I am starting to get the idea. You mean using the script not for creating a physical USB installer stick, but rather for creating an installer image file (to attach it to VM), is that right?

**Comment by mvastola** [comment]
Exactly. So you should be able to do like:
```bash
dd if=/dev/zero of=install-disk.img bs=1G count=8
losetup /dev/loopX install-disk.img
./mkosxinstallusb.sh /dev/loopX install-dir
losetup -d /dev/loopX
```

**Comment by eprigorodov** [comment]
That changes a lot and really makes `kpartx` a universal way of addressing partitions.
I will merge the PR as is and then will fix work with physical USB devices and will add yet another confirmation request before formatting.
Thank you for the good observation, for the fix and for your patience in explanation :)

**Comment by mvastola** [comment]
Ack. Thanks for approving, but I never got to make that change to the regex. 
Can you do it, or should I make another PR? 
Things that need to be done:
-  Change `(\S+)p` in the regex to `([a-z]+)` 
- Anywhere I put `${result_of_map_partition}pX`, the 'p' needs to be deleted.
- The `set +e` should be removed from `map_partitions`
- `set -e` should be added to the top of the script

**Comment by eprigorodov** [comment]
I will do that, thank you.
